### PR TITLE
CBG-1318: Add crc32 xattr hash

### PR DESCRIPTION
**Pull request by JRascagneres**
Added additional CRC32 xattr hash to support import PRs

- [x] http://uberjenkins.sc.couchbase.com:8080/view/All/job/sync-gateway-integration/741/

- [x] Depends on #4964

### CODE REVIEW

File: db/document.go [hunk 1]
```diff
@@ -65,6 +65,7 @@ type SyncData struct {
 	Expiry          *time.Time          `json:"exp,omitempty"`           // Document expiry.  Information only - actual expiry/delete handling is done by bucket storage.  Needs to be pointer for omitempty to work (see https://github.com/golang/go/issues/4357)
 	Cas             string              `json:"cas"`                     // String representation of a cas value, populated via macro expansion
 	Crc32c          string              `json:"value_crc32c"`            // String representation of crc32c hash of doc body, populated via macro expansion
+	Crc32UserXattr  string              `json:"user_xattr_value_crc32"`  // String representation of crc32c hash of user xattr
```
Comment (bbrks) [inline db/document.go]:
```suggestion
	Crc32cUserXattr  string              `json:"user_xattr_value_crc32c"` // String representation of crc32c hash of user xattr
```

**Review by bbrks — CHANGES_REQUESTED** [review]

File: db/document.go (hunk 1 above)
Comment (JRascagneres) [reply db/document.go]: Agree but leaving this for now as we may change this to something like 'meta_crc32c' so this could be expanded in the future.

Comment (adamcfraser) [reply db/document.go]: If we're leaving this, even as a placeholder, I think it makes sense to fix the crc32c in the name.

File: db/database.go [hunk 2]
```diff
@@ -1213,6 +1213,7 @@ func (db *Database) UpdateAllDocChannels(regenerateSequences bool) (int, error)
 						if updatedExpiry != nil {
 							updatedDoc.UpdateExpiry(*updatedExpiry)
 						}
+						doc.SyncData.Crc32UserXattr = base.Crc32cHashString(currentUserXattr)
```
Comment (adamcfraser) [inline db/database.go]: Crc32cHashString() is going to return a non-empty string for a nil currentUserXattr.  I think we want to set doc.SyncData.Crc32cUserXattr to empty string if currentUserXattr isn't present.

File: db/crud.go [hunk 3]
```diff
@@ -1783,6 +1783,7 @@ func (db *Database) updateAndReturnDoc(docid string, allowImport bool, expiry ui
 			deleteDoc = currentRevFromHistory.Deleted
 
 			// Return the new raw document value for the bucket to store.
+			doc.SyncData.Crc32UserXattr = base.Crc32cHashString(userXattr)
```
Comment (adamcfraser) [inline db/crud.go]: Same comment as below, about empty userXattr values.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

**Review by adamcfraser — CHANGES_REQUESTED** [review]
I don't think this has been rebased on CBG-1317 yet, but a couple of comments that I think are independent of that.

File: db/document.go [hunk 4]
```diff
@@ -332,6 +333,19 @@ func (doc *Document) GetMetaMap(userXattrKey string) (map[string]interface{}, er
...
+		doc.SyncData.Crc32cUserXattr = base.Crc32cHashString(doc.rawUserXattr)
+		return
+	}
+
+	// If no key is defined but there is a value in its crc32 then we should update the crc32 to be empty. This would
+	// happen if the feature was enabled but later disabled.
+	if userXattrKey == "" && doc.SyncData.Crc32cUserXattr != "" {
+		doc.SyncData.Crc32cUserXattr = ""
+	}
+}
```
Comment (bbrks) [inline db/document.go]: I think this might be the same comment as Adam's earlier - but if there was a user xattr value on a doc and then it got removed, but the user still has the `userXattrKey` set (e.g. for other docs), won't that keep around the old crc value indefinitely?

I.e. Should we also be emptying the value when `len(rawUserXattr) == 0` ?

**Review by bbrks — CHANGES_REQUESTED** [review]

File: db/document.go [hunk 5]
```diff
@@ -332,6 +333,21 @@ func (doc *Document) GetMetaMap(userXattrKey string) (map[string]interface{}, er
...
+		return
+	}
+
+	// If userXattrKey is still set and there is no rawUserXattr then the xattr has been removed from the doc and we can
+	// set to empty hash.
+	// If userXattr key is no longer set but the crc32 hash is still set then we can set it to empty hash. This likely
+	// means the feature was enabled but later disabled.
+	if len(doc.rawUserXattr) == 0 && userXattrKey != "" || userXattrKey == "" && doc.SyncData.Crc32cUserXattr != "" {
+		doc.SyncData.Crc32cUserXattr = ""
+	}
```
Comment (bbrks) [inline db/document.go]: This is getting a bit hard to follow now... Fairly sure it can be rewritten to minimise the conditionals.

```suggestion
	// Empty the hash if we've no longer got a value or a key defined.
	if len(doc.rawUserXattr) == 0 || userXattrKey == "" {
		doc.SyncData.Crc32cUserXattr = ""
		return
	}

	doc.SyncData.Crc32cUserXattr = base.Crc32cHashString(doc.rawUserXattr)
```

**Review by bbrks — CHANGES_REQUESTED** [review]

File: db/document.go (hunk 5 above)
Comment (JRascagneres) [reply db/document.go]: I considered this but this will mean that even if the user doesn't have, or has never had user xattrs enabled then this property will be added. Not really an issue in terms of space or anything but could be misleading.

Comment (bbrks) [reply db/document.go]: Aren't both of those cases caught by the conditional setting empty string?

Comment (JRascagneres) [reply db/document.go]: Not sure what you mean? Empty userXattrKey means that the feature is disabled. So we'll always set it if its disabled in your suggestion.

Comment (bbrks) [reply db/document.go]: Always set an empty string if the feature is disabled? Is that not what we want?

Comment (JRascagneres) [reply db/document.go]: Well like I say, if the feature has never been enabled it'd be weird to have that as a key in sync data imo.

Comment (bbrks) [reply db/document.go]: That field is a string type. It's always going to be initialized to `""` regardless, and also doesn't have `omitempty` in the JSON struct tag.

File: db/document.go [hunk 6]
```diff
@@ -332,6 +333,21 @@ func (doc *Document) GetMetaMap(userXattrKey string) (map[string]interface{}, er
...
+		return
+	}
+
+	// If userXattrKey is still set and there is no rawUserXattr then the xattr has been removed from the doc and we can
+	// set to empty hash.
+	// If userXattr key is no longer set but the crc32 hash is still set then we can set it to empty hash. This likely
+	// means the feature was enabled but later disabled.
+	if len(doc.rawUserXattr) == 0 || userXattrKey == "" {
+		doc.SyncData.Crc32cUserXattr = ""
+	}
```
Comment (bbrks) [inline db/document.go]: The previous suggestion applied to the whole function, there's no need for another length check if we move the "success" path out and to the bottom, and make the `= ""` case exit.

Comment (JRascagneres) [reply db/document.go]: Yeah makes sense. Sorry flipping between branches and missing things.

File: db/document.go [hunk 7]
```diff
@@ -65,6 +65,7 @@ type SyncData struct {
 	Expiry          *time.Time          `json:"exp,omitempty"`           // Document expiry.  Information only - actual expiry/delete handling is done by bucket storage.  Needs to be pointer for omitempty to work (see https://github.com/golang/go/issues/4357)
 	Cas             string              `json:"cas"`                     // String representation of a cas value, populated via macro expansion
 	Crc32c          string              `json:"value_crc32c"`            // String representation of crc32c hash of doc body, populated via macro expansion
+	Crc32cUserXattr string              `json:"user_xattr_value_crc32c"` // String representation of crc32c hash of user xattr
```
Comment (adamcfraser) [inline db/document.go]: I expected omitempty here - I don't think we need to set the property when empty, correct?

File: db/document.go [hunk 8]
```diff
@@ -332,6 +333,16 @@ func (doc *Document) GetMetaMap(userXattrKey string) (map[string]interface{}, er
 	}, nil
 }
 
+func (doc *Document) SetCrc32cUserXattrHash(userXattrKey string) {
+	// If user xattr is nil or feature is disabled then set hash to ""
+	if len(doc.rawUserXattr) == 0 || userXattrKey == "" {
```
Comment (adamcfraser) [inline db/document.go]: Is there a scenario where userXattrKey == "", but len(doc.rawUserXattr) > 0?  Want to understand the requirement to pass userXattrKey in here.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: db/document.go (hunk 8 above)
Comment (JRascagneres) [reply db/document.go]: I think I've made a mistake when looking at Bens comment above but I think I need to change this back to what it was. I will re-review this.

Comment (bbrks) [reply db/document.go]: If the key is empty, logically the doc doesn't have a user xattr any more (because we don't know where to find one). I think it makes sense to empty the checksum in that case?

Comment (JRascagneres) [reply db/document.go]: Yeah. Ben is right. 
> userXattrKey == "", but len(doc.rawUserXattr) > 0

That bit threw me a little. But yeah that basically checks if the feature is enabled. Its feasible that a customer will enable / disable this and in that case we want to remove any previous uses of the hash.

Comment (adamcfraser) [reply db/document.go]: As discussed, I don't think there's a case where rawUserXattr is non-zero if the userXattrKey is empty.

**Review by bbrks — DISMISSED** [review]

**Review by adamcfraser — APPROVED** [review]
